Issues fixed and new features

BugaIulian opened the pull request.

BugaIulian committed:
Issues fixed and add new features to the app:

-refactor the name of the UserServiceImplements class to WeatherServiceImpl.
-refactor the name of the interface UserService to WeatherService.
-changed the name of the method inside the WeatherService interface from createWeather to storeWeatherObservation. It describes better the action being performed.
-refactor the UserDTO class name to WeatherDTO.
-refactor the UserController class to WeatherController.
-code formatting has been applied to the whole project.
-the name of the "location" String data type, "temperature" double data type
and "humidity" int have been changed to:
* location > city
* temperature > weatherTemperature
* humidity > weatherHumidity
I think it provides more specific information related to our app.
- the 'createWeather' method from the WeatherController class has been renamed to "storeWeather", it is more specific to the role of the app.
-the updateWeather method from the weatherServiceImpl interface has been renamed to getWeatherObservations, it  returns a list of weatherObservationsDTOs.
-add a new method in the WeatherController labelled getWeatherObservations that gets all the Weather DTOs from weatherList DTO's in the WeatherServiceImpl class. This way we can make use of the weatherList created on the mob programming but never had the chance to use it. The endpoint for getWeatherObservations is the same as storeWeather, more exactly "/api/weather".
-add a new method labelled deleteWeatherObservationsByCity in the WeatherService interface, the logic behind it has been implemented in the WeatherService class. The weather controller had also been updated with this method on a new end point "/api/weather/{cityName}".
-add a new method labelled updateWeather in the WeatherService interface, the logic behind it has been implemented in the WeatherService class. The weather controller had also been updated with this method on a new end point "/api/weather".
-logs have been created for each method, we might need to delete them later, however for logging each http request I think it's helpful at this stage.
-the text from the readme.md file has been altered for a better understanding of the edu project.

BugaIulian committed:
Issues fixed and add new features to the app:

-refactor the name of the UserServiceImplements class to WeatherServiceImpl.
-refactor the name of the interface UserService to WeatherService.
-changed the name of the method inside the WeatherService interface from createWeather to storeWeatherObservation. It describes better the action being performed.
-refactor the UserDTO class name to WeatherDTO.
-refactor the UserController class to WeatherController.
-code formatting has been applied to the whole project.
-the name of the "location" String data type, "temperature" double data type
and "humidity" int have been changed to:
* location > city
* temperature > weatherTemperature
* humidity > weatherHumidity
I think it provides more specific information related to our app.
- the 'createWeather' method from the WeatherController class has been renamed to "storeWeather", it is more specific to the role of the app.
-the updateWeather method from the weatherServiceImpl interface has been renamed to getWeatherObservations, it  returns a list of weatherObservationsDTOs.
-add a new method in the WeatherController labelled getWeatherObservations that gets all the Weather DTOs from weatherList DTO's in the WeatherServiceImpl class. This way we can make use of the weatherList created on the mob programming but never had the chance to use it. The endpoint for getWeatherObservations is the same as storeWeather, more exactly "/api/weather".
-add a new method labelled deleteWeatherObservationsByCity in the WeatherService interface, the logic behind it has been implemented in the WeatherService class. The weather controller had also been updated with this method on a new end point "/api/weather/{cityName}".
-add a new method labelled updateWeather in the WeatherService interface, the logic behind it has been implemented in the WeatherService class. The weather controller had also been updated with this method on a new end point "/api/weather".
-logs have been created for each method, we might need to delete them later, however for logging each http request I think it's helpful at this stage.
-the text from the readme.md file has been altered for a better understanding of the edu project.

BugaIulian commented:
@dobraflorin  when you have the time, see how this pull request works for you. Cheers.

AlexandruIliescu commented on src/main/java/org/example/controllers/WeatherController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @PostMapping("/api/weather")
+    public ResponseEntity<WeatherDTO> storeWeather(@RequestBody @Validated WeatherDTO weatherDTO) {
```
If we want to make validations on DTOs, we can use the `@Validated` annotation on the controller class as you did with `WeatherController`, but in the parameter's controller method, we can/should use `@Valid` annotation. Having that in place, we can further use `@NotNull`, `@NotEmpty`, etc annotations on DTOs fields as desired.

You can check [this](https://www.baeldung.com/spring-boot-bean-validation) example. Observe the `@Valid` annotation from the controller class, and annotations from the entity, but in our case, we want to make that validation on the DTO fields.
Please ignore the ExceptionHandler or the testing part from the example.

BugaIulian replied:
Hey Alex, I've added the spring boot starter validation in the pom.xml and validated the fields of the WeatherDTO, yes indeed it seems I got it wrong when it came to the validation on the parameter's methods in the controller class. I have changed it from @Validated to @Valid. The fields from the DTO have been validate with the @NotBlank on the city String DataType and @Max and @Min on the temperature and humidity.

AlexandruIliescu reviewed (approved):
Great job Iulian! Weather APIs are getting amazing improvement on many levels.

BugaIulian committed:
Validation:

-pom.xml has been updated with a new dependency from spring boot framework, more exactly the spring-boot-starter-validation.
-DTO fields have been validated with the @NotBlank constraint, it has been tested and the validation works on the city name.
